python/pyiceberg/expressions/visitors.py

```diff
@@ -675,6 +702,57 @@ def visit_less_than_or_equal(self, term: BoundTerm[L], literal: Literal[L]) -> b
 
         return ROWS_MIGHT_MATCH
 
+    def visit_starts_with(self, term: BoundTerm[L], literal: Literal[L]) -> bool:
+        pos = term.ref().accessor.position
+        field = self.partition_fields[pos]
+        prefix = str(literal.value)
+
+        # contains_null encodes whether at least one partition value is null,
+        # lowerBound is null if all partition values are null
+        all_null = field.contains_null is True and field.lower_bound is None
+        if all_null or not field.lower_bound or not field.upper_bound:
+            return ROWS_CANNOT_MATCH
+
+        lower = _from_byte_buffer(term.ref().field.field_type, field.lower_bound)
+        # truncate lower bound so that its length is not greater than the length of prefix
+        if lower and lower[: len(str(prefix))] > prefix:
+            return ROWS_CANNOT_MATCH
+
+        upper = _from_byte_buffer(term.ref().field.field_type, field.upper_bound)
+        # truncate upper bound so that its length is not greater than the length of prefix
+        if upper and upper[: len(prefix)] < prefix:
+            return ROWS_CANNOT_MATCH
+
+        return ROWS_MIGHT_MATCH
+
+    def visit_not_starts_with(self, term: BoundTerm[L], literal: Literal[L]) -> bool:
+        pos = term.ref().accessor.position
+        field = self.partition_fields[pos]
+        prefix = str(literal.value)
+
+        if field.contains_null or not field.lower_bound or not field.upper_bound:
+            return ROWS_MIGHT_MATCH
+
+        # not_starts_with will match unless all values must start with the prefix. This happens when
+        # the lower and upper bounds both start with the prefix.
+        lower = _from_byte_buffer(term.ref().field.field_type, field.lower_bound)
+        upper = _from_byte_buffer(term.ref().field.field_type, field.upper_bound)
+
+        if lower and upper:
+            # if lower is shorter than the prefix then lower doesn't start with the prefix
+            if len(lower) < len(prefix):
```

Comment:
Same here, should we put the `len(prefix)` in a variable, it looks like we're using it multiple times.